src/Qualify.hs

```diff
@@ -11,147 +27,293 @@ import qualified Set
 import Types
 import Util
 
--- | Changes the symbol part of a defn (the name) to a new symbol path
--- | Example: (defn foo () 123) => (defn GreatModule.foo () 123)
-setFullyQualifiedDefn :: XObj -> SymPath -> XObj
-setFullyQualifiedDefn (XObj (Lst [defn, XObj _ symi symt, args, body]) i t) newPath =
-  XObj (Lst [defn, XObj (Sym newPath Symbol) symi symt, args, body]) i t
-setFullyQualifiedDefn (XObj (Lst [def, XObj _ symi symt, expr]) i t) newPath =
-  XObj (Lst [def, XObj (Sym newPath Symbol) symi symt, expr]) i t
-setFullyQualifiedDefn xobj _ = error ("Can't set new path on " ++ show xobj)
+--------------------------------------------------------------------------------
+-- Errors
+
+-- | Error qualifying a symbol.
+data QualificationError
+  = FailedToQualifyDeclarationName XObj
+  | FailedToQualifySymbols XObj
+  | FailedToQualifyPath SymPath
+
+instance Show QualificationError where
+  show (FailedToQualifyDeclarationName xobj) =
+    "Couldn't fully qualify the definition: " ++ pretty xobj
+  show (FailedToQualifySymbols xobj) =
+    "Couldn't fully qualify the symbols in the form: " ++ pretty xobj
+  show (FailedToQualifyPath spath) =
+    "Couldn't fully qualify the symbol: " ++ show spath
+      ++ "in the given context."
+
+--------------------------------------------------------------------------------
+-- Data
+
+-- | Denotes an XObj containing only symbols that *have been fully qualified*.
+--
+-- A fully qualified xobj **must not** be qualified further (e.g. using context
+-- paths).
+newtype Qualified = Qualified {runQualified :: XObj}
```

Comment:
To me, `run` signals that there's a monadic action going on.

Reply:
Good point! changed.

Reply:
3cdc04e